src/System Application/App/Agent/Setup/AccessControls/ViewAgentPermissions.Page.al

```diff
@@ -133,29 +140,60 @@ page 4334 "View Agent Permissions"
             PermissionSetNotFound := not AggregatePermissionSet.Get(Rec.Scope, Rec."App ID", Rec."Role ID");
 
         if not ShowCompanyFieldOverride then begin
-            ShowCompanyField := not AccessControlForSingleCompany(GlobalSingleCompanyName);
+            ShowCompanyField := not AgentImpl.TryGetAccessControlForSingleCompany(Rec."User Security ID", GlobalSingleCompanyName);
             CurrPage.Update(false);
         end;
     end;
 
-    local procedure AccessControlForSingleCompany(var SingleCompanyName: Text[30]): Boolean
+    local procedure CopyAccessControlToBuffer(UserSecurityID: Guid; var TempAccessControlBuffer: Record "Access Control Buffer" temporary)
     var
-        TempCompany: Record Company temporary;
-        UserSettings: Codeunit "User Settings";
+        AccessControl: Record "Access Control";
     begin
-        UserSettings.GetAllowedCompaniesForUser(Rec."User Security ID", TempCompany);
-        if TempCompany.Count() <> 1 then
-            exit(false);
+        TempAccessControlBuffer.Reset();
+        TempAccessControlBuffer.DeleteAll();
+
+        AccessControl.SetRange("User Security ID", UserSecurityID);
+        if not AccessControl.FindSet() then
+            exit;
+
+        repeat
+            Clear(TempAccessControlBuffer);
+            TempAccessControlBuffer."Company Name" := AccessControl."Company Name";
+            TempAccessControlBuffer.Scope := AccessControl.Scope;
+            TempAccessControlBuffer."App ID" := AccessControl."App ID";
+            TempAccessControlBuffer."Role ID" := AccessControl."Role ID";
+            TempAccessControlBuffer.Insert();
+        until AccessControl.Next() = 0;
+    end;
 
-        SingleCompanyName := TempCompany.Name;
-        exit(true);
+    local procedure SaveAccessControlBuffer(UserSecurityID: Guid; var TempAccessControlBuffer: Record "Access Control Buffer" temporary)
+    var
+        AccessControl: Record "Access Control";
+    begin
+        // Delete all existing access control records for the agent
+        AccessControl.SetRange("User Security ID", UserSecurityID);
```

Comment:
I'm not sure that this is the best thing to do. Maybe we can follow the same pattern of deleting row by row to avoid locking out some scenarios. Row by row update would be less disrupting. 
The flow would be the same, first delete the sets that do not exist or insert new ones. There are few records here so we can afford spending a bit more time on this